File: 2023/05/haskell/05.hs
{- stack script --resolver lts-14.20 
                --package split -}
module Main where

import Data.List.Split
import Data.List

type Map = [Int]
type Mapper = [Map]

parseSeeds :: String -> [Int]
parseSeeds s = map read $ tail $ splitOn " " s

parseMappers :: [String] -> [Mapper]
parseMappers ss = map (map (map read . splitOn " ") . tail) [s | s <- splitOn [""] ss, not $ null s]

parse :: [String] -> ([Int], [Mapper])
parse (s:ss) = (parseSeeds s, parseMappers ss)

apply :: Mapper -> Int -> Int
apply []     x = x
apply (m:ms) x | x' /= x   = x'
               | otherwise = apply ms x
  where
    x' = applyMap m x

applyMap :: Map -> Int -> Int
applyMap [a, b, c] x | b <= x && x < b + c = x - b + a
                     | otherwise           = x

task1 :: [Int] -> [Mapper] -> Int
task1 ss ms = minimum $ task1' ss ms
  where
    task1' :: [Int] -> [Mapper] -> [Int]
    task1' ss    []  = ss
    task1' ss (m:ms) = flip task1' ms $ flip map ss $ apply m

apply2 :: Mapper -> (Int, Int) -> [(Int, Int)]
apply2 []     x = [x]
apply2 (m:ms) x | snd x'' /= snd x = x'
                | otherwise        = init x' ++ apply2 ms x''
  where
    x'  = applyMap2 m x
    x'' = last x'

applyMap2 :: Map -> (Int, Int) -> [(Int, Int)]
applyMap2 [a, b, c] (l, r) = [(x, y) | (x, y) <- [(l, min r $ b - 1), (a - b + max l b, a - b + (min r $ b + c - 1)), (max l $ b + c, r)], x <= y]

task2 :: [Int] -> [Mapper] -> Int
task2 ss ms = fst $ minimum $ task2' ss' ms
  where
    ss' = pairs ss

    pairs :: [Int] -> [(Int, Int)]
    pairs []       = []
    pairs (x:y:xs) = (x, x + y - 1) : pairs xs

task2' :: [(Int, Int)] -> [Mapper] -> [(Int, Int)]
task2' ss    []  = ss
task2' ss (m:ms) = flip task2' ms $ sort $ flip concatMap ss $ apply2 m

main :: IO ()
main = do
  input <- flip fmap getContents lines
  let (seeds, mappers) = parse input
  putStr "Task 1: "
  let mappers' = flip map mappers $ sortBy (\[a, b, c] [d, e, f] -> compare [b, a, c] [e, d, f])
  print $ task1 seeds mappers'
  putStr "Task 2: "
  print $ task2 seeds mappers'

File: 2023/05/rust/05.rs
use std::io;
use std::cmp;

type Map = Vec<i64>;
type Mapper = Vec<Map>;

fn task1(seeds: &Vec<i64>, mappers: &Vec<Mapper>) -> i64 {
  let mut seeds = seeds.clone();
  for mapper in mappers {
    for seed in &mut seeds {
      for map in mapper {
        if map[1] <= *seed && *seed < map[1] + map[2] {
          *seed += map[0] - map[1];
          break;
        }
      }
    }
  }
  *seeds.iter().min().unwrap()
}

fn task2(seeds: &Vec<i64>, mappers: &Vec<Mapper>) -> i64 {
  let mut pairs = Vec::new();
  for i in (0..seeds.len()).step_by(2) {
    pairs.push((seeds[i], seeds[i] + seeds[i + 1] - 1));
  }
  pairs.sort();
  for mapper in mappers {
    pairs = pairs.iter().map(|pair| {
      let mut new_pairs = Vec::new();
      let mut current = pair.clone();
      for map in mapper {
        if current.0 < map[1] {
          new_pairs.push((current.0, cmp::min(current.1, map[1] - 1)));
          current.0 = map[1];
          if current.0 > current.1 {
            break;
          }
        }
        if current.0 < map[1] + map[2] {
          new_pairs.push((cmp::max(current.0, map[1]) + map[0] - map[1], cmp::min(current.1, map[1] + map[2] - 1) + map[0] - map[1]));
          current.0 = map[1] + map[2];
          if current.0 > current.1 {
            break;
          }
        }
      }
      if current.0 <= current.1 {
        new_pairs.push((current.0, current.1));
      }
      new_pairs
    }).flatten().collect::<Vec<(i64, i64)>>();
    pairs.sort();
  }
  pairs.iter().fold(i64::MAX, |acc, pair| cmp::min(acc, pair.0))
}

fn main() {
  let mut line = String::new();

  io::stdin().read_line(&mut line).unwrap();
  let mut seed_split = line.trim().split(" ");
  seed_split.next();
  let seeds = seed_split.map(|x| x.parse::<i64>().unwrap()).collect::<Vec<i64>>();
  line.clear();
  io::stdin().read_line(&mut line).unwrap();
  line.clear();

  let mut mappers = Vec::new();
  let mut mapper = Vec::new();
  while io::stdin().read_line(&mut line).unwrap() > 0 {
    if line.trim() == "" {
      mapper.sort_by(|a: &Map, b: &Map| a[1].cmp(&b[1]));
      mappers.push(mapper);
      mapper = Vec::new();
    }
    else if (line.as_bytes()[0] as char).is_digit(10) {
      let map = line.trim().split(" ").map(|x| x.parse::<i64>().unwrap()).collect::<Vec<i64>>();
      mapper.push(map);
    }
    line.clear();
  }
  mapper.sort_by(|a: &Map, b: &Map| a[1].cmp(&b[1]));
  mappers.push(mapper);
  println!("Task 1: {}", task1(&seeds, &mappers));
  println!("Task 2: {}", task2(&seeds, &mappers));
}
